src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationReleaseOrderService.cs

```diff
@@ -0,0 +1,226 @@
+            }
+
+            var releaseOrder = publication.ReleaseOrders.Find(ro => ro.ReleaseId == releaseId)
+                ?? throw new KeyNotFoundException($"No matching ReleaseOrder found for {nameof(Release)} with ID {releaseId}");
+
+            if (releaseOrder.IsAmendment)
+            {
+                var releaseAmendment = publication.Releases.Find(r => r.Id == releaseId)
+                    ?? throw new KeyNotFoundException($"No matching amendment for {nameof(Release)} with ID {releaseId} found");
+
+                var originalReleaseOrder = publication.ReleaseOrders.Find(r => r.ReleaseId == releaseAmendment.PreviousVersionId)
+                    ?? throw new KeyNotFoundException($"No matching ReleaseOrder for original {nameof(Release)} with ID {releaseAmendment.PreviousVersionId} found");
+
+                publication.ReleaseOrders.Remove(originalReleaseOrder);
+
+                releaseOrder.IsAmendment = false;
+                ResetOrders(publication.ReleaseOrders);
+            }
+
+            releaseOrder.IsDraft = false;
+
+            _context.Update(publication);
+        }
+
+        public async Task UpdateForUpdateCombinedReleaseOrder(
+            Guid publicationId,
+            List<CombinedReleaseUpdateOrderViewModel> releaseOrderUpdates)
+        {
+            var publication = await GetPublication(publicationId);
+
+            foreach (var update in releaseOrderUpdates)
+            {
+                if (update.Id == Guid.Empty)
+                {
+                    throw new ApplicationException($"Update with order {update.Order} must have a release ID specified");
+                }
+
+                if (update.Order == 0)
+                {
+                    throw new ApplicationException($"Updated order for release with ID {update.Id} must be greater than 0");
+                }
+
+                if (update.IsAmendment)
+                {
+                    var releaseAmendment = publication.Releases.Find(r => r.Id == update.Id)
+                        ?? throw new KeyNotFoundException($"No matching amendment for {nameof(Release)} with ID {update.Id} found");
+
+                    var originalReleaseOrder = publication.ReleaseOrders.Find(ro => ro.ReleaseId == releaseAmendment.PreviousVersionId)
+                        ?? throw new KeyNotFoundException($"No matching ReleaseOrder for original {nameof(Release)} with ID {releaseAmendment.PreviousVersionId} found");
+
+                    originalReleaseOrder.Order = update.Order;
+                }
+
+                var releaseOrder = publication.ReleaseOrders.Find(ro => ro.ReleaseId == update.Id)
+                    ?? throw new KeyNotFoundException($"No matching ReleaseOrder found for {(update.IsLegacy ? nameof(LegacyRelease) : nameof(Release))} with ID {update.Id}");
+
+                releaseOrder.Order = update.Order;
+
+                _context.Update(publication);
+            }
+        }
```

Comment:
I think this is quite complex and I would have tried to avoid any attributes like `IsDraft` which are stateful from being in the ReleaseOrder's.

We should be able to find out from the release itself what state it is currently in rather than relying on `IsDraft`.

A 'draft' release in this service means unpublished whereas draft in release terminology is just one of the unpublished states that a release can be in so that's also slightly confusing.

Duplicating the state adds the possibility that it could become out of sync and we see that there's a change required to the Publisher function to maintain this state which has added to the complexity of the Publisher, as well as put a dependency on the Admin project. 

`IsDraft` is also not applicable to legacy releases so it's not so nice that it appears in every persisted entry.

`ReleaseOrder.IsLegacy` could be avoided by giving `ReleaseOrder` separate id attributes like `Guid? ReleaseId` and `Guid? LegacyReleaseId`. There can then be a computed property like `public bool IsLegacy => LegacyReleaseId.HasValue` if necessary.

Assuming that ReleaseOrder needs no other attributes and that null values get omitted when serialising to Json, I imagine it's possible to get this working with the persisted value looking something like this:

```
[{"ReleaseId":"76447fab-ca79-4918-acde-08dc3394e05a"},
 {"ReleaseId":"5f047886-4193-40c8-acdf-08dc3394e05a"},
 {"LegacyReleaseId":"02ea926b-c95a-4e80-81a3-8035adaea191"}]
```

compared to what you have now which looks like this

```
[{"ReleaseId":"76447fab-ca79-4918-acde-08dc3394e05a","IsLegacy":false,"IsDraft":true,"IsAmendment":false,"Order":1},
 {"ReleaseId":"5f047886-4193-40c8-acdf-08dc3394e05a","IsLegacy":false,"IsDraft":true,"IsAmendment":false,"Order":2},
 {"ReleaseId":"02ea926b-c95a-4e80-81a3-8035adaea191","IsLegacy":true,"IsDraft":false,"IsAmendment":false,"Order":3}]
```

To me, we shouldn't really care whether a release is draft or an amendment. Say we've got an ordered list of release id's for a publication. Then in the content API when producing the public view model we get the list of published releases, can't we simply order that list of published releases against the order they are observed in the Publication? It won't matter that the Publication's list of ordered releases has draft or amended release id's in it as they will be ignored.

To make this simpler, you could consider swapping the release id's for `ReleaseParent` id's.

All releases derived from the same initial version have the same `ReleaseParent`. It never changes when amending a release. In future I intend to rename `ReleaseParent -> Release` and `Release -> ReleaseVersion`.

If you use `ReleaseParent` as the id to track the order:

- You won't need to ever delete an entry because they will always exist once the first version is created. The first version can never be removed (for now) so a parent exists forever.
- You won't need to do anything on publishing to update draft state.
- You won't need to do create new entries in the amendment service when a new release version is created.

Any remaining logic could be done elsewhere and you might not need this service at all.

Reply:
Ben said: `ReleaseOrder.IsLegacy could be avoided by giving ReleaseOrder separate id attributes like Guid? ReleaseId and Guid? LegacyReleaseId. There can then be a computed property like public bool IsLegacy => LegacyReleaseId.HasValue if necessary.`

This is a larger change and perhaps should be a different ticket, but I was thinking there is even a possibility to remove the `LegacyReleases` table entirely, since they're just a description and a URL, and so could be stored in the `ReleaseOrders` JSON - no ID is required, the order is implied by it's place in the JSON array, and it's in the `Publications` table so no need to keep a `PublicationId` either.

This would be a big simplification and remove a huge amount of code, because then practically all of `LegacyReleaseController` could be deleted.

I think it's definitely worth exploring how much effort this would be now, as I think it would be a big win. But also happy to push this to another ticket if it's too much.